### PR TITLE
Add support for complex intersection/union filters

### PR DESCRIPTION
**Pull request by V-FEXrt**
Improves/increases the types of queries that can be spelled out with the db inspection subsystem.

The syntax is as follows
  - repeating a flag is intersection
  - `,` in a flag unions the parts (`,` can be repeated for multiple parts)
  - `*` matches 0 or more  of any character
  - `?` matches exactly 1 of any character

Example
`wake --job "5??,6??" --label "*c++*" --label "*.c" --output "*.o"` is

 - AND
   - label contains `c++`
   - label ends with `.c`
   - job output a file ending in `.o`
   - OR
     - job id is 3 digits starting with 5
     - job id is 3 digits starting with 6

### CODE REVIEW

File: tools/wake/cli_options.h [hunk 1]
```diff
@@ -115,10 +118,10 @@ struct CommandLineOptions {
       {'C', "chdir", GOPT_ARGUMENT_REQUIRED},
       {0, "in", GOPT_ARGUMENT_REQUIRED},
       {'x', "exec", GOPT_ARGUMENT_REQUIRED},
-      {0, "job", GOPT_ARGUMENT_REQUIRED},
+      {0, "job", GOPT_ARGUMENT_REQUIRED | GOPT_REPEATABLE_VALUE, job_ids_buffer.data(), (unsigned int)argc_in},
```
Comment (JakeSiFive) [inline tools/wake/cli_options.h:121]: I hate how this is done uhg. oh well

File: tools/wake/main.cpp [hunk 2]
```diff
@@ -117,9 +117,21 @@ std::string globish_to_like(const std::string &str) {
...
-  std::string glob(str);
-  return globish_to_like(glob);
+template <class QueryF>
+Set<long> apply_inspection_query(std::unordered_map<long, JobReflection> &captured_jobs,
+                                 Set<long> current,
+                                 const std::vector<std::vector<std::string>> &query, QueryF qf) {
+  for (const std::vector<std::string> &and_part : query) {
+    std::vector<JobReflection> hits = {};
+    for (const std::string &or_part : and_part) {
+      if (or_part == "") continue;
```
Comment (JakeSiFive) [inline tools/wake/main.cpp:127]: Why is this here?

Comment (V-FEXrt) [reply tools/wake/main.cpp:127]: small optimization that allows for trailing comma.

`wake --job "1,2,3,"`  splits into `["1", "2", "3", ""]`

Comment (JakeSiFive) [reply tools/wake/main.cpp:127]: But now you can't match on the empty string because of it. It changes the semantics.

Comment (JakeSiFive) [reply tools/wake/main.cpp:127]: Use case: `wake --label ""` to find all the jobs that people did not add labels to

Comment (V-FEXrt) [reply tools/wake/main.cpp:127]: hmm fair. I don't feel strongly towards this and it does change the semantics

**Comment by V-FEXrt** [comment]
Ope, I thought I was merging the other PR. @JakeSiFive can PTAL/approve if you are happy. If not I'll revert

**Review by JakeSiFive — COMMENTED** [review]
Yeah this looks good to me
